storage/lmdb/src/lib.rs

```diff
@@ -0,0 +1,199 @@
+// Copyright (c) 2022 RBB S.r.l
+// opensource@mintlayer.org
+// SPDX-License-Identifier: MIT
+// Licensed under the MIT License;
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://github.com/mintlayer/mintlayer-core/blob/master/LICENSE
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+mod error;
+
+use std::path::PathBuf;
+
+use lmdb::Cursor;
+use storage_core::{
+    backend::{self, TransactionalRo, TransactionalRw},
+    info::{DbDesc, MapDesc},
+    Data, DbIndex,
+};
+use utils::sync::Arc;
+
+/// Metadata about the list of databases (key-value maps)
+#[derive(Eq, PartialEq, Debug, Clone)]
+struct DbList(Vec<lmdb::Database>);
```

Comment:
What do you mean by array here? `Box<[lmdb::Database]>` or something else?

The user-facing interface enforces quite a few invariants and also guarantees the indices it passes to the backend are in range. The backend itself does not create or manipulate indices explicitly. I don't think adding strong types here has a good ratio of benefits vs. added complexity.

Reply:
The backend does not have access to the schema as a type parameter, so it does not know the size at compilation time. It would be possible to pass it in but not sure it's worth it and I suspect it will be too rigid when implementing e.g. database migrations.

`Box<[lmdb::Database]>` is still dynamically sized, just without some of the `Vec` methods that allow you to modify the list such as `.push()`.

Reply:
The way I solved this problem before was that the Database enum came from the common database interface. The set/get methods would take that enum, which is type-safe, then for every implementation of the database a `const fn` would do the translation to an integer, which makes it even possible to use an array of maps instead of map of maps. 

In our case, this can be a separate module or even crate if we want to maintain perfect core that doesn't know anything about the application layer. You don't have to do this right now though. You can keep this as a future thing to do since it's not specifically related to lmdb.